In [3]:
import numpy as np
import pandas as pd
import sweetviz as sv
import pandas as pd
from sklearn.model_selection import train_test_split
from supervised.automl import AutoML
from tabulate import tabulate

# Grupo

Renato Aurélio Fernandes Totti - 171586

Thiago Pereira Correa - 171085

André Chierighini - 171340

Ciro Guimarães - 152539

Camila de Carvalho Mendes - 171660

Letícia Vigna Modenese Silva - 171869

---


# Leitura dos dados;

Primeiro é analizado os dados brutos:

In [6]:


df_dataset = pd.read_csv('./data/creditcard.csv', sep=',', index_col=None) 


print(df_dataset.shape)


df_dataset.describe()




(284807, 31)


Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.654067e-16 -3.568593e-16  2.578648e-16  4.473266e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.340915e-16  1.683437e-15 -3.660091e-16 -1.227390e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

In [7]:
df_dataset.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

# Análise exploratória;

Aqui pode-se perceber que como os dadosa v1 ~ v28 são dados anonimizados das transações é difícil encontrar valores condizentes, os dados de Time e Amount foram removidos por apresentar valores discrepantes quando comparados aos outros

In [ ]:
df_datasetWithouAmountAndTime = df_dataset
df_datasetWithouAmountAndTime =df_dataset.drop(['Time','Amount'], axis = 1)
df_datasetWithouAmountAndTime.plot.box(figsize=(15,7))


Box plot de tempo mostra um valor adequado já que é a contagem de tempo desde o início

In [ ]:
df_dataset['Time'].plot.box()

In [ ]:
df_dataset['Amount'].plot.box()

Após toda analise inicial, é feita a primeira limpeza onde é removido duplicados, depois gerado uma nova analise com base no atributo "Class", o qual indica qual linha é fraude ou não.

Aqui utiliza-se a biblioteca chamada sweetviz, que gera um relatório com as informações mais relevantes para a decisão da utilização dos dados do dataset separando os dados que mais se correlacionam com o `target_feat`, nesse caso `Class`

In [ ]:
import sweetviz as sv
df_dataset = pd.read_csv('./data/creditcard.csv', sep=',', index_col=None) 
df_dataset=df_dataset.drop_duplicates()
my_report = sv.analyze(df_dataset,target_feat='Class')
my_report.show_notebook()

Um segundo relátorio gerado com o sweetivz, agora com os dados separados, demonstra a distribuição das classes de fraudes dentro do data set comparada ás transações legítimas.

In [ ]:
df_dataset = pd.read_csv('./data/creditcard.csv', sep=',', index_col=None) 
df_dataset=df_dataset.drop_duplicates()
df_dataset_fraude = df_dataset.loc[df_dataset['Class']==1]
df_dataset_legitmo = df_dataset.loc[df_dataset['Class']==0]
my_report = sv.compare([df_dataset_legitmo,'Legítimo'],[df_dataset_fraude,'Fraude'])
my_report.show_notebook()

É necessário pontuar que como a maioria das colunas são dados anonimizados é de dificil leitura como por exemplo, os dados clássicos da íris setosa, pois apresentam extrema variação por nao se tratar de um dataset "orgânico" e sim um dataset de amostras coletadas de comportamento humano, sendo assim, a análise exploratória nos diz que esse data set pode contem transações fraudulentas de extrema dificuldade de classificação caso elas se assemelhem ao comportamento de uma pessoa realizando uma transação legítima.

# Pré processamento de dados

Feita a analise a partir da classe foi feita uma seleção das colunas que tem uma maior relação no que diz sobre a linha ser fraude ou não, pela matriz de correlação. O sweetviz também facilita a decisão das colunas mais relevantes junto a aba de Class.

Vale notar que as colunas e linhas já estavam apropriadas ao uso desse programa, portanto não foi necessária a limpeza profunda. Os relatórios gerados demonstram que nenhum dado estava faltante.

Para verificar que a escolha de colunas nao afetou os dados é necessário gerar outro report.

In [ ]:
df_dataset = pd.read_csv('./data/creditcard.csv', sep=',', index_col=None) 
df_dataset=df_dataset[['Amount','V17','V14','V12','V10','V16','V3','V7','V11','V4','V18','V1','Class']]
my_report = sv.analyze(df_dataset,target_feat='Class')
my_report.show_notebook()

É possível perceber que a redução de colunas gerou linhas com dados idênticos, portanto é importante remover dados duplicatos

In [ ]:
df_dataset = pd.read_csv('./data/creditcard.csv', sep=',', index_col=None) 
df_dataset=df_dataset[['Amount','V17','V14','V12','V10','V16','V3','V7','V11','V4','V18','V1','Class']]
#removendo dados duplicatos
df_dataset=df_dataset.drop_duplicates()
my_report = sv.analyze(df_dataset,target_feat='Class')
my_report.show_notebook()

# Construção e validação do modelo;
Agora com os dados limpos, é feito o treinamento de varios modelos diferentes e após isso é gerado a pontuação(score) de cada modelo de teste treinado.

A validação utilizada foi a validação cruzada com k=10, para manter o rigor estatístico e evitar bias dos modelos gerados

Para treino foi utilizado um split de 75% e 25% para testes.

In [32]:

modelos=['Random Forest','Xgboost','Neural Network','Decision Tree','CatBoost']

df_dataset = pd.read_csv('./data/creditcard.csv', sep=',', index_col=None) 
df_dataset=df_dataset[['Amount','V17','V14','V12','V10','V16','V3','V7','V11','V4','V18','V1','Class']]
#removendo dados duplicatos
df_dataset=df_dataset.drop_duplicates()

X_train, X_test, y_train, y_test = train_test_split(
    df_dataset[df_dataset.columns[:-1]], df_dataset["Class"], test_size=0.25
)
for modelo in modelos:
    automl = AutoML(results_path=f'AC2_{modelo}',algorithms=[modelo],eval_metric='f1',explain_level=1,validation_strategy={"validation_type": "kfold","k_folds": 10, "shuffle": True,"random_seed": 1997},total_time_limit=60*30)
    automl.fit(X_train, y_train)



This model has already been fitted. You can use predict methods or select a new 'results_path' for a new 'fit()'.
This model has already been fitted. You can use predict methods or select a new 'results_path' for a new 'fit()'.
This model has already been fitted. You can use predict methods or select a new 'results_path' for a new 'fit()'.
This model has already been fitted. You can use predict methods or select a new 'results_path' for a new 'fit()'.
This model has already been fitted. You can use predict methods or select a new 'results_path' for a new 'fit()'.


---

# Discussão dos resultados.

Para entendermos os modelos primeiramente é necessário saber quantas fraudes existem no dataset de teste

In [34]:
totFrauds=len(y_test.loc[y_test==1])
print(f'Fraudes contidas no set de testes:{totFrauds}')

Fraudes contidas no set de testes:103


Também é preciso calcular a score de cada modelo perante ao set de testes

In [35]:
for modelo in modelos:
    automl = AutoML(results_path=f'AC2_{modelo}')
    score = automl.score(X_test,y_test)
    print(f"Modelo:{modelo} score:{round(score*100,2)}%")

Modelo:Random Forest score:99.94%
Modelo:Xgboost score:99.97%
Modelo:Neural Network score:99.94%
Modelo:Decision Tree score:99.93%
Modelo:CatBoost score:99.96%


In [43]:
modelos=['Random Forest','Xgboost','Neural Network','Decision Tree','CatBoost']
modelNumber=0

errorsHigh=[0,0,0,0,0]
dataHigh=[[],[],[],[],[]]

errorslow=[0,0,0,0,0]
dataLow=[[],[],[],[],[]]
for modelo in modelos:
    automl = AutoML(results_path=f'AC2_{modelo}')
    pred=automl.predict(X_test)
    i=0
    errors=0
    for y in y_test:
        if(y==1 and pred[i]==0):
            amt=X_test.loc[i,'Amount']
            if(amt >= 100):
                dataHigh[modelNumber].append(amt)
                errorsHigh[modelNumber]+=1
            else:
                dataLow[modelNumber].append(amt)
                errorslow[modelNumber]+=1
            errors+=1
        i+=1
    modelNumber+=1

    print(f'Modelo:{modelo} | Falsos negativos:{errors} | Porcentagem de falsos negativos:{round((errors/totFrauds)*100,2)}%')


Modelo:Random Forest | Falsos negativos:25 | Porcentagem de falsos negativos:24.27%
Modelo:Xgboost | Falsos negativos:18 | Porcentagem de falsos negativos:17.48%
Modelo:Neural Network | Falsos negativos:20 | Porcentagem de falsos negativos:19.42%
Modelo:Decision Tree | Falsos negativos:29 | Porcentagem de falsos negativos:28.16%
Modelo:CatBoost | Falsos negativos:18 | Porcentagem de falsos negativos:17.48%


É possível perceber que os falsos negativos giram em torno de 20-28% de todas as predições porém esse valor sozinho nao quer dizer toda a análise necessária. É muito interessante analisar os valores desse falsos positivos.

Foi escolhido um valor arbitrário de 100 euros para representar fraudes de valor considerável

In [44]:
print('Tabela de erros com mais de 100 euros')

#transposição da tabela para formatar o print
dataHigh = pd.DataFrame(dataHigh).T
print(tabulate(dataHigh,headers=modelos))

Tabela de erros com mais de 100 euros
      Random Forest    Xgboost    Neural Network    Decision Tree    CatBoost
--  ---------------  ---------  ----------------  ---------------  ----------
 0          1096.99    1096.99           1096.99          1096.99     1096.99
 1           519.9      519.9             519.9            105.89      519.9
 2           208.58     208.58            208.58           519.9       208.58
 3           129        129               129              208.58      129
 4           104.81     104.81            104.81           129         104.81
 5           634.3      634.3             634.3            104.81      634.3
 6           247.86     105.99            105.99           634.3       105.99
 7           105.99     237.26            311.91          1809.68      237.26
 8           237.26     311.91            nan              105.99      311.91
 9           311.91     nan               nan              311.91      nan


Como é possível perceber, o modelo de árvore de decisão teve o menor desempenho contendo o maior classificando uma fraude de 1800 euros como uma transação legítima. 
Porém as várias vezes que o set de teste foi mudado a floresta randômica tambémfoi sucetivel a erros do tipo.


É interessante notar como alguns dos valores se repetem dentro dos vários modelos, evidenciando uma possivel similaridade com transações legitimas. 

In [38]:
print('Tabela de erros com menos de 100 euros')

#transposição da tabela para formatar o print
dataLow = pd.DataFrame(dataLow).T
print(tabulate(dataLow,headers=modelos))

Tabela de erros com menos de 100 euros
      Random Forest    Xgboost    Neural Network    Decision Tree    CatBoost
--  ---------------  ---------  ----------------  ---------------  ----------
 0             1          1                 1                2           1
 1             1          1                 1                1           1
 2             1          8                 1                1           8
 3             8          2.47              8                1           0
 4             0          3.14              0                8           2.47
 5             2.47       0.76              2.47             0           3.14
 6            25          3.76              3.14             2.47        0.76
 7             3.14      42.53              0.76            25          11.39
 8             0.76     nan                 3.76             3.14        3.76
 9             1        nan                 0.68             0.76        0.68
10             3.76     nan          

## Matrizes de confusão e outras métricas

## CatBoost


## Metric details
|           |      score |     threshold |
|:----------|-----------:|--------------:|
| logloss   | 0.00309474 | nan           |
| auc       | 0.96851    | nan           |
| f1        | 0.849612   |   0.474383    |
| accuracy  | 0.999531   |   0.474383    |
| precision | 0.931973   |   0.474383    |
| recall    | 1          |   1.81519e-05 |
| mcc       | 0.852726   |   0.474383    |


## Confusion matrix (at threshold=0.474383)
|                     |   Predicted as negative |   Predicted as positive |
|:--------------------|------------------------:|------------------------:|
| Labeled as negative |                  206376 |                      20 |
| Labeled as positive |                      77 |                     274 |

---

## Decision Tree


## Metric details
|           |      score |   threshold |
|:----------|-----------:|------------:|
| logloss   | 0.00381487 |  nan        |
| auc       | 0.889383   |  nan        |
| f1        | 0.779874   |    0.443866 |
| accuracy  | 0.999323   |    0.443866 |
| precision | 0.852234   |    0.443866 |
| recall    | 0.994203   |    0        |
| mcc       | 0.782371   |    0.443866 |


## Confusion matrix (at threshold=0.443866)
|                     |   Predicted as negative |   Predicted as positive |
|:--------------------|------------------------:|------------------------:|
| Labeled as negative |                  206359 |                      43 |
| Labeled as positive |                      97 |                     248 |

---

## Neural Network


## Metric details
|           |      score |     threshold |
|:----------|-----------:|--------------:|
| logloss   | 0.00386043 | nan           |
| auc       | 0.950215   | nan           |
| f1        | 0.822653   |   0.452456    |
| accuracy  | 0.999424   |   0.452456    |
| precision | 0.846626   |   0.452456    |
| recall    | 1          |   3.8731e-153 |
| mcc       | 0.822696   |   0.452456    |


## Confusion matrix (at threshold=0.452456)
|                     |   Predicted as negative |   Predicted as positive |
|:--------------------|------------------------:|------------------------:|
| Labeled as negative |                  206352 |                      50 |
| Labeled as positive |                      69 |                     276 |

---

## Random Forest

## Metric details
|           |      score |     threshold |
|:----------|-----------:|--------------:|
| logloss   | 0.00337719 | nan           |
| auc       | 0.914741   | nan           |
| f1        | 0.799387   |   0.459631    |
| accuracy  | 0.999366   |   0.459631    |
| precision | 0.847403   |   0.459631    |
| recall    | 1          |   0.000208493 |
| mcc       | 0.800361   |   0.459631    |


## Confusion matrix (at threshold=0.459631)
|                     |   Predicted as negative |   Predicted as positive |
|:--------------------|------------------------:|------------------------:|
| Labeled as negative |                  206355 |                      47 |
| Labeled as positive |                      84 |                     261 |

---

## Extreme Gradient Boosting (Xgboost)

## Metric details
|           |    score |   threshold |
|:----------|---------:|------------:|
| logloss   | 0.133929 | nan         |
| auc       | 0.921456 | nan         |
| f1        | 0.853168 |   0.456902  |
| accuracy  | 0.999541 |   0.456902  |
| precision | 0.913907 |   0.456902  |
| recall    | 1        |   0.0130897 |
| mcc       | 0.854835 |   0.456902  |


## Confusion matrix (at threshold=0.456902)
|                     |   Predicted as negative |   Predicted as positive |
|:--------------------|------------------------:|------------------------:|
| Labeled as negative |                  206376 |                      26 |
| Labeled as positive |                      69 |                     276 |

---

É interessante notar as altas taxas de precisão e acurácia para os modelos principalmente os modelos de xgboost catboost e rede neural. Porém é possivel notar a alta taxa de falsos negativos dentro da matriz de confusão, como citado no capítulo de análise exploratória, fraudes que se comportam de modo homólogo á uma transação legítima é extremamente difícil discernir as duas.

# Conclusão

O Extreme Grandient Boost (xgboost) sempre se mostrou o melhor modelo quanto a acurária (em raras vezes igual ou menor ao CatBoost). Totdavia a rede neural sempre teve taxa de sucesso em não criar falsos negativos com transações de alto valor. Por mais que a acurácia dela seja menor, provavelmente se msotra como a mais interessante a ser aplicada para a resolução do problema.

Vale citar que os falsos negativos apresentados pelos modelos quase sempre são relacionados as mesmas transações evidenciando a dificuldade em modelar e classificar comportamento humano dentro de um problema de IA.